capsules/extra/src/date_time.rs

```diff
@@ -0,0 +1,354 @@
+            7 => Ok(date_time::Month::July),
+            8 => Ok(date_time::Month::August),
+            9 => Ok(date_time::Month::September),
+            10 => Ok(date_time::Month::October),
+            11 => Ok(date_time::Month::November),
+            12 => Ok(date_time::Month::December),
+            _ => Err(ErrorCode::INVAL),
+        }
+    }
+
+    fn dotw_as_u32(&self, dotw: date_time::DayOfWeek) -> u32 {
+        match dotw {
+            date_time::DayOfWeek::Sunday => 0,
+            date_time::DayOfWeek::Monday => 1,
+            date_time::DayOfWeek::Tuesday => 2,
+            date_time::DayOfWeek::Wednesday => 3,
+            date_time::DayOfWeek::Thursday => 4,
+            date_time::DayOfWeek::Friday => 5,
+            date_time::DayOfWeek::Saturday => 6,
+        }
+    }
+
+    fn u32_as_dotw(&self, dotw_num: u32) -> Result<date_time::DayOfWeek, ErrorCode> {
+        match dotw_num {
+            0 => Ok(date_time::DayOfWeek::Sunday),
+            1 => Ok(date_time::DayOfWeek::Monday),
+            2 => Ok(date_time::DayOfWeek::Tuesday),
+            3 => Ok(date_time::DayOfWeek::Wednesday),
+            4 => Ok(date_time::DayOfWeek::Thursday),
+            5 => Ok(date_time::DayOfWeek::Friday),
+            6 => Ok(date_time::DayOfWeek::Saturday),
+            _ => Err(ErrorCode::INVAL),
+        }
+    }
+
+    /// Transforms two u32 numbers into a DateTimeValues structure (year, month, dotm, dotw, hour, minute, seconds)
+    /// Check file documentation for details on how the u32 tuple stores data
+
+    fn date_from_u32_tuple(
+        &self,
+        date: u32,
+        time: u32,
+    ) -> Result<date_time::DateTimeValues, ErrorCode> {
+        let month_num = date % (1 << 9) / (1 << 5);
+        let month_name = self.u32_as_month(month_num)?;
+
+        let dotw_num = time % (1 << 20) / (1 << 17);
+        let dotw_name = self.u32_as_dotw(dotw_num)?;
+
+        let date_result = date_time::DateTimeValues {
+            year: (date % (1 << 21) / (1 << 9)) as u16,
+            month: month_name,
+            day: (date % (1 << 5)) as u8,
+
+            day_of_week: dotw_name,
+            hour: (time % (1 << 17) / (1 << 12)) as u8,
+            minute: (time % (1 << 12) / (1 << 6)) as u8,
+            seconds: (time % (1 << 6)) as u8,
+        };
+
+        if !(date_result.day <= 31) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.hour <= 24) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.minute <= 60) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.seconds <= 60) {
+            return Err(ErrorCode::INVAL);
+        }
+
+        Ok(date_result)
+    }
+
+    /// Transforms DateTimeValues structure (year, month, dotm, dotw, hour, minute, seconds) into two u32 numbers
+    /// Check file documentation for details on how the u32 numbers stores data
+    /// The two u32 numbers are returned as a tuple
+
```

Comment:
```suggestion
```

Reply:
Why do these keep getting resolved? The empty line is here and needs to be removed.